_posts/2018-05-02-load-balanced-smart-proxies-with-haproxy.md

```diff
@@ -0,0 +1,138 @@
+| AMQP       | 5647 | TCP  | `roundrobin` | port 5647 on all proxies |
+| Puppet     | 8140 | TCP  | `roundrobin` | port 8140 on all proxies |
+| PuppetCA   | 8141 | TCP  | `roundrobin` | port 8140 on the first proxy |
+| SmartProxy | 9090 | TCP  | `roundrobin` | port 9090 on all proxies |
+| Docker     | 5000 | TCP  | `roundrobin` | port 5000 on all proxies |
+
+[You can find an example `haproxy.cfg` in the `forklift` repository.](https://github.com/theforeman/forklift/blob/master/roles/haproxy/templates/haproxy.cfg.j2)
+
+## Configuring Clients
+
+As far as clients are concerned, they should only know about `smartproxy.example.test`, and should never need to connect directly to a smartproxy. That has to be configured for Puppet, goferd, and for RHSM/Yum.
+
+### Using katello-client-bootstrap
+
+[katello-client-bootstrap](https://github.com/Katello/katello-client-bootstrap) is designed to configure a machine to be used with an exiting Katello environment. It also can configure Puppet properly for us, and supports [setting a separate Puppet CA and Puppet CA port in current Git](https://github.com/Katello/katello-client-bootstrap/pull/250).
+
+To attach clients to the load-balanced proxy, we just have to pass `--server smartproxy.example.test` and `--puppet-ca-port 8141`, `bootstrap.py` will take care of the rest.
+
+### Manual registration
+
+Instead of using `bootstrap.py`, the client can also be registered manually. However, we recommend using `bootstrap.py` since it handles these extra manual steps for you.
+
+After installing `katello-ca-consumer` RPM from `http://smartproxy.example.test/pub/katello-ca-consumer-latest.noarch.rpm` `/etc/rhsm/rhsm.conf` and `/etc/yum.repos.d/redhat.repo` will contain the hostname of whichever proxy served the initial request and not `smartproxy.example.net` as we'd like it. Thus, when registering the system with `subscription-manager`, we have to pass `--serverurl=https://smartproxy.example.test:8443/rhsm --baseurl=https://smartproxy.example.test/pulp/repos` to update those two config files.
+
+Also Puppet needs to be configured to talk to the split-out PuppetCA, so the `puppet.conf` needs the following lines:
+
+```
+server = smartproxy.example.test
+ca_server = smartproxy.example.test
+ca_port = 8141
+```
+
+## Forklift
+As usual, there is [support for the loadbalanced scenario in `forklift`](https://github.com/theforeman/forklift/blob/master/pipelines/pipeline_katello_lbproxy.yml).
+
+## Help! Something doesn't work!
+The load balancer adds additional complexity to a Foreman/Katello installation. If a client is having trouble using one of the services provided by the smart proxies, a typical first debugging step is to point the client directly to one of the smart proxies via a `/etc/hosts` entry, which bypasses the load balancer. If the issue persists, then you know that the issue is with that smart proxy. Each smart proxy can be checked individually for functionality with this method.
```

Comment:
Is it worth nothing that pointing the client directly to one of the smart proxies will be subject to restrictions if the smart proxy they point it too is not the first one?

Reply:
@beav this paragraph was initially written by you, so what do you think?

I'd add something like "As the PuppetCA functionality is only available on the first smart proxy, it can only be tested when Puppet is configured to use port 8140 on the first smart proxy, or port 8141 on the load-balancer for PuppetCA requests."

Reply:
For me that addition sounds reasonable.